### DISCUSSION TITLE
Providing composer dependencies to Rules

### DISCUSSION THREAD

**Discussion by peterfox** [rectorphp/rector#7955] [Ideas]
Sometimes it would be better if rules could be aware of the composer.json or composer.lock file of a project it's being used on.

For example, when I created the [assertStatusToAssertMethod rule for Laravel](https://github.com/driftingly/rector-laravel/blob/main/src/Rector/MethodCall/AssertStatusToAssertMethodRector.php), the rule is now slightly problematic because the possible conversions that occur as specific to the version of the framework that is installed. In some instances, the rule will convert the code to use a method that isn't available.

While some of this could be configured via options to specify which methods are available, it seems better to just provide a mechanism for rules to know what versions of PHP or packages are installed within the project.

An example of what this should be like is the [Composer Runtime utilities](https://getcomposer.org/doc/07-runtime.md).

```php
\Composer\InstalledVersions::isInstalled('vendor/package'); // returns bool
\Composer\InstalledVersions::isInstalled('psr/log-implementation');

\Composer\InstalledVersions::satisfies(new VersionParser, 'vendor/package', '2.0.*');
\Composer\InstalledVersions::satisfies(new VersionParser, 'psr/log-implementation', '^1.0');
```

**Comment by TomasVotruba** [comment]
Hi, we would love this!

Could you send a PR with proposal?

**Comment by samsonasik** [comment]
The `Composer\InstalledVersions` may be prefixed, then after that, it read prefixed class:

https://github.com/rectorphp/rector/blob/9b4ae51505fb73319145df86e2a2c69c512387b8/vendor/composer/InstalledVersions.php#L12C1-L25

That may not be able to read "root" installed composer packages?

**Comment by peterfox** [comment]
@TomasVotruba I'd definitely like to have a go at it, but I can't promise anything. Time-wise, I'm quite slammed. I have a rough idea for how it could work at least.

**Reply by samsonasik** [reply]
@peterfox just ensure to test it directly in vendor/ first to avoid possible issue I commented at above https://github.com/rectorphp/rector/discussions/7955#discussioncomment-6021029

**Comment by peterfox** [comment]
I've looked at this more recently. I can make this work but I'd need to tie into some kind of event bus to swap the composer dependencies during the execution of the rule and then restore Rector's composer dependencies after the rule has executed.

**Reply by TomasVotruba** [reply]
I think it's much easier to kick it off. We already apply similar mechanism to polyfill packages for PHP upgrade sets.

**Reply by peterfox** [reply]
Do you have an example @TomasVotruba ?

**Reply by TomasVotruba** [reply]
This one: https://github.com/rectorphp/rector-src/pull/5388
